ui/helpers/utils/tags.ts

```diff
@@ -0,0 +1,41 @@
+// TODO: Remove restricted import
+// eslint-disable-next-line import/no-restricted-paths
+import { getEnvironmentType } from '../../../app/scripts/lib/util';
+import { getIsUnlocked } from '../../ducks/metamask/metamask';
+import {
+  ApprovalsMetaMaskState,
+  getInternalAccounts,
+  getPendingApprovals,
+  getTransactions,
+  selectAllTokensFlat,
+} from '../../selectors';
+import { getMetamaskNotifications } from '../../selectors/metamask-notifications/metamask-notifications';
+import { NftState, selectAllNftsFlat } from '../../selectors/nft';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function getStartupTraceTags(state: any) {
```

Comment:
The use of `any` here is against our contributor guidelines, and the later type casts are advised against as well. I understand that situation we're in regarding types for the UI Redux state has no great solutions though, not without a lot of work that would be out of scope here at least.

What I would recommend here in place of `any` is that we use `unknown` or `Json`. And for the three selectors that require a more specific type, we can `// @ts-expect-error` those for now until we get correct types for this state.

I know our guidelines recommend that we use "correct types" first (e.g. type guards and `satisfies` and so on), then type assertions second, then escape hatches like `ts-expect-error` third, so you might think it's preferable to use a type assertion here. But that's only when the type you're casting to is correct, and I don't think it's correct in any of these cases. So IMO it's better to use an escape hatch for now.

Reply:
Very much agree with Mark on this. Here are some initial suggestions for correct types here.
I also added property checks since some of the selectors lack runtime validation.

```diff
diff --git a/ui/helpers/utils/tags.ts b/ui/helpers/utils/tags.ts
index d274e7575e..a67edbc1b4 100644
--- a/ui/helpers/utils/tags.ts
+++ b/ui/helpers/utils/tags.ts
@@ -1,9 +1,9 @@
 // TODO: Remove restricted import
 // eslint-disable-next-line import/no-restricted-paths
+import { ApprovalControllerState } from '@metamask/approval-controller';
 import { getEnvironmentType } from '../../../app/scripts/lib/util';
 import { getIsUnlocked } from '../../ducks/metamask/metamask';
 import {
-  ApprovalsMetaMaskState,
   getInternalAccounts,
   getPendingApprovals,
   getTransactions,
@@ -11,20 +11,42 @@ import {
 } from '../../selectors';
 import { getMetamaskNotifications } from '../../selectors/metamask-notifications/metamask-notifications';
 import { NftState, selectAllNftsFlat } from '../../selectors/nft';
+import { NotificationServicesController } from '@metamask/notification-services-controller';
+import { TransactionControllerState } from '@metamask/transaction-controller';
+import { AccountsControllerState } from '@metamask/accounts-controller';
+import { KeyringControllerState } from '@metamask/keyring-controller';
+import { TokensControllerState } from '@metamask/assets-controllers';
 
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-export function getStartupTraceTags(state: any) {
+export function getStartupTraceTags(state: {
+  metamask: Partial<
+    KeyringControllerState &
+      AccountsControllerState &
+      TransactionControllerState &
+      TokensControllerState
+  > &
+    ApprovalControllerState &
+    NftState['metamask'] &
+    NotificationServicesController.NotificationServicesControllerState;
+}) {
   const uiType = getEnvironmentType();
   const unlocked = getIsUnlocked(state) as boolean;
-  const accountCount = getInternalAccounts(state).length;
-  const nftCount = selectAllNftsFlat(state as unknown as NftState).length;
-  const notificationCount = getMetamaskNotifications(state).length;
-  const tokenCount = selectAllTokensFlat(state).length as number;
+  const accountCount =
+    'internalAccounts' in state.metamask
+      ? getInternalAccounts(state).length
+      : 0;
+  const nftCount =
+    'allNfts' in state.metamask ? selectAllNftsFlat(state).length : 0;
+  const notificationCount =
+    'metamaskNotificationsList' in state.metamask
+      ? getMetamaskNotifications(state).length
+      : 0;
+  const tokenCount =
+    'allTokens' in state.metamask
+      ? (selectAllTokensFlat(state).length as number)
+      : 0;
   const transactionCount = getTransactions(state).length;
 
-  const pendingApprovals = getPendingApprovals(
-    state as unknown as ApprovalsMetaMaskState,
-  );
+  const pendingApprovals = getPendingApprovals(state);
 
   const firstApprovalType = pendingApprovals?.[0]?.type;
```

Reply:
Switched to `MetaMaskReduxState`.